### DISCUSSION TITLE
passing data in custom components

### DISCUSSION THREAD

**Discussion by cadentic** [vercel/next.js#40166] [Help] [answered]
hi

if I am creating a new custom component in ../component/Testimonial.js  which holds material UI components. 

eg. 

`const Testimonial =(props) => {
return (
    <Testimonial>
            <Toolbar>

                {/* {Pages.data.map(Pages => ( */}

                  {/*  <div classname={style.main}> */}

                    <Card sx={{ flexDirection: 'row', maxWidth: 545, borderRadius: "20px", spacing: "125px", boxShadow: "7px", alignItems: 'stretch' }} elevation={4} position="relative">

                        <CardHeader

                            avatar={

                                <Avatar sx={{ bgcolor: red[200], spacing: "5px", margin: "2px" }} aria-label="Technocrat">

                                    TC

                                </Avatar>

                            }

                            action={

                                <IconButton aria-label="settings">

                                    <MoreVertIcon />

                                </IconButton>

                            }

                            title={props.Title}

                            subheader="September 14, 2016"

                        />

                        <CardMedia

                            component="img"

                            height="200"

                            image="{props.Images}"

                            alt="Paella dish"

                        />

                        <CardContent>

                            <Typography variant="body2" color="text.secondary">

                                {props.Description}</Typography>

                        </CardContent>

                        <CardActions enableSpacing>

                            <IconButton aria-label="add to favorites">

                                <FavoriteIcon />

                                <Rating name="half-rating" defaultValue={2.5} precision={0.5} />

                            </IconButton>

                            <IconButton aria-label="share">

                                <ShareIcon />

                            </IconButton>

                            <ExpandMore

                                expand={expanded}

                                onClick={handleExpandClick}

                                aria-expanded={expanded}

                                aria-label="show more"

                            >

                                <ExpandMoreIcon />

                            </ExpandMore>

                        </CardActions>

                        <Collapse in={expanded} timeout="auto" unmountOnExit>

                            <Typography paragraph padding="10px">

                                <Box sx={{ my: 8, spacing: 10, margin: 2 }}>  {props.Content} </Box></Typography>

                        </Collapse>

                    </Card>

                    {/*     </div> */}

   )) </Toolbar>

        {/* </ div > */}

        {/* </ div > */}

        </Testimonial>

    );

};

`

how can i pass these data while planting them from other pages?
i need to re-use those components multiple times on multiple pages. to do so we need to pass the data for each `props`

**Comment by icyJoseph** [comment]
I am not sure I understand, but you could use a solution such as SWR or React-Query and then access them from within the components. 

Alternatively something like Redux, Valtio, Jotai, etc, could be used to construct global state, which you can consume from within this component. 

Here's a write up by a team member of Vercel, which I think is relevant, https://github.com/vercel/swr/discussions/1408#discussioncomment-1383367.

If you think I've misunderstood you, perhaps it'd help to draw things out? Maybe you could use https://excalidraw.com/ and share what you are thinking about graphically.

**Comment by cadentic** [comment]
say i have some jsx files eg .../components/Testi.js 
it contains : 
```
import * as react from react 

const Testimonial = ({ Pages, error }) => { 

 return (
<> 
<Toolbar >
{Pages.map(Pages => (
<H1> Pages.attributes.Content</H1>
))} </Toolbar>
</Toolbar>

</>

);
};



export default Testimonial;
```
it will throw an error 
```
from pages/index.js


const Home = ({ Pages, error }) => {
}
return (
 <Testimonial data={{Pages,error}} /> 

  );

}

export default Home;

export const getInitialProps = async ctx => {
  try {

    const res = await axios.get('http://localhost:1337/api/' + 'pages');
    const Pages = res.data;

    return { Pages };
  } catch (error) {
    return { error };
  }
};
```




simply I just need to pass this data to the component.

**Comment by icyJoseph** [comment, marked as answer]
A few pointers on your snippet:

- `getInitialProps` must be attached to the page, for example, `Home.getInitialProps`
- I think you might want to use `getServerSideProps`, but then your function has to return `return {props: {Pages}}`
-  It is usually not a good idea to try and reach for `api/routes` from within pages. Read https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes.
- The snippet has an error how Home is defined:
```jsx
const Home = ({ Pages, error }) => {
} // -> This ends the Home Component and keeps the return floating
return (
 <Testimonial data={{Pages,error}} /> 

  );

}

export default Home;
```

Here's a couple of examples:

#### with getInitialProps:

```tsx
import Link from "next/link";

const PageComponent = ({ Pages }) => {
  return (
    <>
      {Pages.map((p) => (
        <span key={p}>{p}</span>
      ))}
    </>
  );
};

const Home = ({ Pages }) => {
  console.log({ Pages });
  return (
    <div>
      Home <Link href="/blog">blog</Link>
      <div>more stuff</div>
      <PageComponent Pages={Pages} />
    </div>
  );
};

Home.getInitialProps = async (ctx) => {
  return { Pages: [1, 2, 3] };
};

export default Home;
```

#### with getServerSideProps:

```tsx
import Link from "next/link";

const PageComponent = ({ Pages }) => {
  return (
    <>
      {Pages.map((p) => (
        <span key={p}>{p}</span>
      ))}
    </>
  );
};

const Home = ({ Pages }) => {
  console.log({ Pages });
  return (
    <div>
      Home <Link href="/blog">blog</Link>
      <div>more stuff</div>
      <PageComponent Pages={Pages} />
    </div>
  );
};

export const getServerSideProps = async (ctx) => {
  return { props: { Pages: [1, 2, 3] } }; // Notice that there's a props key in this case!
}; 

export default Home;
```

**Reply by cadentic** [reply]
still i have an error like @icyJoseph 
![Capture_link](https://user-images.githubusercontent.com/55697529/188910998-57484e7f-26af-4bbc-a1ca-485f5d5cb6da.JPG)

**Reply by icyJoseph** [reply]
That means that `Pages` is undefined. 

Let's try to analyse why.

Your code, as you have shown, has two blocks inside the data fetching function, a try block and a catch block.

It is likely that the try block is running into an exception and the catch block is then triggered, which means that your Pages props would be, indeed, undefined.

If the try block is running correctly, then are you passing the props correctly to the child component?

### Possible lead

Please do understand that at this point this is all React and JavaScript code that you got to understand and be able to debug. Do you have a repository I could take a look at? **I found this on your repositories**:

```js
// https://github.com/cadentic/strapi-webdev/blob/main/zettadaten_fe_next17/pages/index.js
const Home = ({ Pages, error }) => {
// }
 // const [Pages, setPages] = useState()
  // setPages(await res.json())

  return (
    <div>
      
      <Navigations />
     {/*  <TabbedSection /> */}
      <div className={style.main}>
      <HeroSection />
       < DropDownPaperSection />
       {/* <Testimonial data={{Pages,error}} /> */}
      </div>
      <CopyerightZettadaten />
    </div>
  );
}


export default Home;

// ❌❌❌❌ This is wrong ❌❌❌
export const getInitialProps = async ctx => {
  try {

    const res = await axios.get('http://localhost:1337/api/' + 'pages');
    const Pages = res.data;
    console.log(Pages);

    return { Pages };
  } catch (error) {
    return { error };
  }
};

// do this instead
Home.getInitialProps = async ctx => {
  try {

    const res = await axios.get('http://localhost:1337/api/' + 'pages');
    const Pages = res.data;
    console.log(Pages);

    return { Pages };
  } catch (error) {
    return { error };
  }
};

// or keep the definition separate
const getInitialProps = async ctx => {
  try {

    const res = await axios.get('http://localhost:1337/api/' + 'pages');
    const Pages = res.data;
    console.log(Pages);

    return { Pages };
  } catch (error) {
    return { error };
  }
};
// and attach it like this to the Home page
Home.getInitialProps = getInitialProps;
```

And I see that you are defining `getInitialProps`, but you never attach it to page, Home in this case. Before `getServerSideProps`, and `getStaticProps` came to be, `getInitialProps` was the way to fetch data on the server, and instead of exporting it, it was attached to the page object. See this page, https://nextjs.org/docs/api-reference/data-fetching/get-initial-props, in the docs for more details.

**Reply by cadentic** [reply]
now it's worked like a charm. thx @icyJoseph
